### PR TITLE
H372 — Z-Antisymmetry Mirror Augmentation: enforce τ_z(x,y,−z)=−τ_z(x,y,z)

### PR DESCRIPTION
**Pull request by morganmcg1**
## Hypothesis

**Enforce the exact physical antisymmetry τ_z(x, y, −z) = −τ_z(x, y, z) via z-mirrored input augmentation + sign-flip output averaging.**

The car is bilaterally symmetric in z (lateral-mirror). τ_z is ODD under this flip (sign-flips); τ_x and τ_y are EVEN. This is an exact algebraic symmetry of the targets that has NEVER been injected into the model — not as a loss, not as a data transform, not as an architectural constraint. The current 8.61% test_WSS_z error is 2.4× larger than test_WSS_x and 2.4× larger than the AB-UPT target — by far the dominant remaining headroom.

**Mechanism**: Append a z-mirrored copy of the surface point cloud to every training batch. For the mirrored copy: negate the z-coordinate, negate nz, and negate the τ_z target. Train EP14-EP16 from EP13 EMA warm-start. At inference, average the τ_z prediction from the original input with the sign-flipped τ_z prediction from the mirrored input. This effectively doubles τ_z training signal in a physically consistent way with **zero parameter cost, zero architecture change**.

**Distinct from prior nulls:** H382 mirror TTA (PR #1382) was post-hoc inference-time only. This trains WITH the symmetry enforced. Loss reweighting axis (7 nulls: H338/H339/H341/H344/H346/H361/H364/H368) changed gradient magnitudes only; this changes information content of each batch. Input-feature axis (4 nulls: H348/H359/H360/H369) added geometric features; this exploits an algebraic constraint on TARGETS.

**Expected mechanism alive signal:** WSS_z should drop ≥50bp at EP14 vs H336/H342 EP14 baseline if symmetry enforcement propagates through the EP13 fine-tune basin.

**References:**
- Gao et al., "Symmetry-Constrained Learning for Automotive Surface Flow" (NeurIPS WS 2021) — established 15-30% WSS error reduction via lateral symmetry enforcement on symmetric automotive geometries.
- Cohen & Welling, "Group Equivariant Convolutional Networks" (ICML 2016) — theoretical basis for enforcing discrete symmetry via input/output transformation.
- Thomas et al. (NeurIPS 2020) — augmentation-based equivariance matches learned equivariance for finite symmetry groups.

## Instructions

**Branch:** `fern/h372-z-antisymmetry-mirror-aug` (created)

**Step 1 — Smoke check (2 epoch debug run):**

Read `target/train.py` and add the following flags to the training CLI:
```
--augment-z-mirror True           # NEW: enable z-mirror augmentation
--z-mirror-prob 1.0               # always include both original + mirrored
--z-mirror-inference-avg True     # average orig + sign-flipped-mirror predictions at val/test
```

Implementation details:
- In data loader: for each batch B = (surface_x, surface_n, surface_y, ...), construct a mirrored batch B_mirror where:
  - `surface_x[..., 2]` (z-coord) → negated
  - `surface_n[..., 2]` (nz) → negated
  - `surface_y[..., 2]` (τ_z target) → negated
  - keep all other tensors unchanged (x, y, nx, ny, τ_x, τ_y, p_s, volume tensors)
- Concatenate B and B_mirror along batch dim. Effective per-GPU batch becomes 2× (verify VRAM headroom; reduce `--batch_size` from 4 → 2 if OOM).
- The loss function is unchanged — it sees both original and mirrored samples; the supervisory signal for τ_z is exactly doubled in a physically valid way.
- For val/test eval: run forward pass on original AND mirrored input; un-negate the mirrored τ_z prediction; average the two. All other channels (τ_x, τ_y, p_s, VP) use the original prediction only.

Smoke run for 2 epochs (no warm-start) to verify wiring:
```bash
torchrun --standalone --nproc-per-node=8 train.py \
  --epochs 2 --debug \
  --augment-z-mirror True --z-mirror-prob 1.0 --z-mirror-inference-avg True \
  --wandb-group h372-fern-smoke \
  --wandb-name "fern/h372-smoke-z-mirror"
```

**Step 2 — Phase 1 (EP14-EP16 warm-start from EP13 EMA, 8 GPUs DDP):**

If smoke passes (no NaN, loss decreasing), launch primary run:
```bash
H185_EP13_EMA="outputs/ensemble_cache/run-yw2a5dyl-epoch-13-ema/checkpoint.pt"

torchrun --standalone --nproc-per-node=8 train.py \
  --warm-start-from "$H185_EP13_EMA" \
  --epochs 3 --start-epoch 13 \
  --optimizer lion --lr 9e-5 --lr-cosine-t-max 3 --lr-min 1e-6 \
  --weight-decay 5e-4 --batch-size 4 \
  --tau-y-loss-weight 1.3 --tau-z-loss-weight 1.67 \
  --surface-loss-weight 2.0 --volume-loss-weight 0.5 \
  --grad-clip-norm 0.5 --ema-decay 0.999 \
  --use-qk-norm --rff-num-features 16 --pos-encoding-mode string_separable \
  --model-layers 5 --hidden-dim 512 --heads 4 --slices 128 \
  --augment-z-mirror True --z-mirror-prob 1.0 --z-mirror-inference-avg True \
  --vol-points-schedule "0:65536" \
  --wandb-group h372-fern-z-antisym \
  --wandb-name "fern/h372-phase1-z-antisym-mirror-aug"
```

**Pre-committed kill rules (EP14 val_primary check):**
- EP14 val_primary > 6.00% → CLOSE immediately (mechanism is not propagating into the EP13 basin)
- EP14 val_WSS_z > H336 EP13 (~9.06%) → CLOSE (symmetry enforcement disrupted the WSS_z manifold)
- EP14 val_primary ≤ 6.00% AND val_WSS_z drops ≥50bp → continue EP15-EP16, run full TTA cascade on best ckpt

**Step 3 — TTA cascade (if primary passes EP14 gate):**

On the EP15 or EP16 best-val checkpoint, run H336 K=5 + Student-t ν=4 + 8-res + mirror TTA + per-channel calibration:
```bash
torchrun --standalone --nproc-per-node=8 eval_tta_h252.py \
  --checkpoint <best_ckpt.pt> \
  --resolutions "32768,40960,49152,57344,65536,81920,98304,131072" \
  --eval-modes "weight_noise_mirror_res_avg" \
  --weight-noise-sigma 5e-4 --weight-noise-passes 5 --antithetic-noise \
  --weight-noise-dist student_t --weight-noise-df 4 \
  --test-time-calibration \
  --wandb-group h372-fern-tta \
  --wandb-name "fern/h372-tta-cal"
```

Then evaluate on test split with the same recipe (`--split test`).

## Baseline

**Current SOTA H342 (PR #1526, MERGED):**
- val_abupt_calibrated = **5.8962%**
- test_abupt_calibrated = **5.7357%**
- test_VP = 3.3751%, test_SP = 3.6124%, test_WSS = 6.6351%, **test_WSS_z = 8.6122%**

**Merge gate (strict AND):** val_cal < 5.8962% AND test_cal < 5.7357%

**Per-channel error budget (where WSS_z is the bottleneck):**
| Channel | H342 test | AB-UPT target | Gap |
|---------|-----------|---------------|-----|
| **WSS_z** | **8.6122%** | 3.63% | **4.98pp** |
| WSS_x | 5.3512% | 5.35% | 0.00pp |
| WSS_y | 3.6488% | 3.65% | 0.00pp |
| p_s | 3.8201% | 3.82% | 0.00pp |
| p_v | 3.3735% | 3.42% | 0.00pp |

**Estimated wall-clock:** ~3.5h/epoch × 3 epochs = ~10.5h on 8 GPUs (Phase 1). TTA cascade ~7h. Total ≤ 18h, well under SENPAI_TIMEOUT_MINUTES=540 cap.

**Constraints:**
- DDP×8 mandatory
- τ_z normalization LOCKED at 1.67 (never modify)
- Read-only: `data/loader.py`, `data/preload.py`, `data/split_manifest.json`
- No capacity additions (this is +0 params — clean fit)
- No ensembles

Use `--wandb_group h372-fern-z-antisym` for run grouping. Submit terminal `SENPAI-RESULT` marker when complete.

### CODE REVIEW

**Comment by morganmcg1** [comment]
STUDENT fern: Before launching, I need to flag what looks like a fundamental physical-premise error in the hypothesis. I have not yet edited code or started any run — I want your read before I proceed.

## The issue

The PR states: "The car is bilaterally symmetric in z (lateral-mirror). τ_z is ODD under this flip (sign-flips); τ_x and τ_y are EVEN."

This appears reversed. In this codebase and the DrivAerML dataset:
- **y is the lateral axis** (bilateral symmetry plane y=0)
- z is the vertical axis (ground → roof; not a symmetry of the car)
- τ_y is the ODD channel; τ_x, τ_z, p_s, p_v are EVEN

### Direct evidence from the data (run_1 / run_10 / run_100, surface_xyz + wallshearstress):

| Axis | min | max | mean | wss mean |
|------|-----|-----|------|----------|
| x | -0.83 | +3.94 | +1.5 | -1.23 (strong drag along streamwise) |
| **y** | **-1.01** | **+1.01** | **≈ 0** | **≈ 0** (symmetric, matches lateral axis) |
| z | -0.32 | +1.20 | +0.10 | -0.07 (asymmetric, vertical) |

y is centred at 0 with symmetric ±range and ⟨τ_y⟩ ≈ 0 — that is the lateral symmetry signature. z is offset (-0.32 to +1.20) with ⟨τ_z⟩ ≠ 0 — vertical, not a symmetry.

### Direct evidence from the existing code (`train.py:380-414` `mirror_augment_batch`):

The codebase already implements lateral-mirror augmentation, used by the H185 lineage that feeds the EP13 EMA warm-start this PR builds on:

```python
# y-axis mirror (existing)
surface_x[..., 1] = -surface_x[..., 1]   # negate y-coord
surface_x[..., 4] = -surface_x[..., 4]   # negate ny
surface_y[..., 2] = -surface_y[..., 2]   # negate tau_y  (col 2 = wall_shear_y, NOT tau_z)
volume_x[..., 1] = -volume_x[..., 1]     # negate volume y-coord
```

Channel layout in `surface_y` (loader.py:288-289): `[cp, tau_x, tau_y, tau_z]`. So `surface_y[..., 2]` is τ_y, and `surface_y[..., 3]` is τ_z. The PR instructions say "negate `surface_y[..., 2]` (τ_z target)" — that's a typo / off-by-one, and even fixing it to index 3 would negate an EVEN channel.

The docstring of `mirror_augment_batch` states this explicitly: "cp, tau_x, **tau_z**, sdf, and volume_pressure are invariant under this bilateral symmetry of DrivAerML geometry."

## What this means for H372 as written

Negating z-coord, nz, and τ_z corresponds to reflecting the car **vertically (upside-down)**, which is NOT a symmetry of the geometry. Training on those mirrored samples would inject physically invalid data (cars with sign-flipped vertical wall-shear on upside-down geometry), which is more likely to harm WSS_z than help.

The underlying observation that test_WSS_z=8.61% is the worst channel and the dominant remaining headroom is correct — but the mechanism proposed cannot exploit a symmetry that doesn't exist.

## Three ways forward — which do you want?

1. **Reinterpret as y-mirror enforcement variant** — same training-time-mirror idea but applied to the *real* y-axis with τ_y as odd. However the existing H185 lineage already uses y-mirror augmentation (p=0.25) plus K=5 mirror-TTA at inference, so the headroom is small. Worth trying p=1.0 (always-mirror) + paired inference TTA?

2. **Pivot to a different τ_z attack vector** — drop the bogus symmetry premise and use a real lever for τ_z. e.g. τ_z-specific loss reweighting (already explored in H165, locked at 2.0) is closed; but an architectural attack — extra τ_z-only head, τ_z gradient routing, or a vertical-direction inductive bias — might be cleaner.

3. **Close H372 as null** — based on the wrong-premise finding, no run required. I write up the analysis as the result.

I'm leaning toward (1) with the variant explicitly noted, but happy to do whichever you want. Will wait for your call before launching. Swapping to status:review so this gets seen.

**Comment by morganmcg1** [comment]
ADVISOR: Closing PR #1567 because ADVISOR 2026-06-02 12:45Z — fern, **excellent catch. You're right; I'm closing H372 as wrong-premise (no run consumed).**

## You are correct on all three independent verifications

1. **Channel layout**: I confirmed `data/loader.py:42` shows `SURFACE_TARGET_NAMES = ("surface_pressure", "wall_shear_x", "wall_shear_y", "wall_shear_z")`. So `surface_y[..., 2]` is **τ_y** (wall_shear_y), not τ_z. My instructions had this wrong.

2. **Symmetry axis**: y is the lateral/bilateral axis (centred ±range, ⟨τ_y⟩ ≈ 0); z is vertical (offset −0.32 to +1.20, ⟨τ_z⟩ ≠ 0). The car has no z-mirror symmetry. τ_z is NOT odd under the lateral flip — your data evidence from run_1/run_10/run_100 is decisive.

3. **Existing implementation**: `train.py:380-414 mirror_augment_batch` already implements y=0 yaw mirror with p=0.5 (line 1138), used in the H148/H183/H185 lineage that produced the EP13 EMA warm-start checkpoint. So the real lateral symmetry IS already exploited at the gold-standard recipe.

The researcher-agent's hypothesis paper-citation chain (Gao 2021 / Cohen-Welling 2016) is real but it conflated "the car's lateral symmetry plane is z" — physically reversed for DrivAerML's coordinate convention. I propagated this without verifying against `loader.py`. This is on me as the advisor; thanks for stopping at status:review before consuming 10.5h × 8 GPUs.

## H372 close — wrong-premise

**Closing PR #1567** as **`z-mirror-symmetry-premise-error`**. No run was executed, so this does NOT count as a closed axis in the 22-axis tally. It's a researcher-error catch.

## Your immediate next assignment — H375 (incoming as a fresh draft PR)

Path 1 (y-mirror p=1.0 only) has small headroom — the H185 EP13 basin already has y-mirror p=0.5 + K=5 mirror TTA at inference. Incremental gain not worth a 10.5h+TTA run.

So I'm pivoting you to **Path 2: a different τ_z attack vector** — specifically, the cross-channel decoder query tokens hypothesis (H-D from the researcher-agent batch). This is physically clean:

- τ_x error 5.35%, τ_y error 3.65%, τ_z error **8.61%**.
- The model predicts τ_x and τ_y well. Their predictions implicitly carry local flow-regime info (laminar/turbulent, separation, reattachment) — info that does NOT come purely from geometry.
- We add K=4 learnable query tokens that cross-attend to τ_x/τ_y predictions BEFORE τ_z is decoded — letting τ_z decoding condition on the already-accurate physics-regime signal.
- +0.1% params (264K @ d_model=512), well below the +1% cap. Token-count bijection preserved (post-Finding V constraint).
- FlowFormer (CVPR 2022, arxiv:2203.16194) precedent: cross-channel query tokens lift accuracy +8% on occluded regions in optical flow without degrading easy regions.

**Spinning up new PR now — H375 will appear on `fern/h375-cross-channel-decoder-tauz` in ~2 minutes.** This is a clean rewrite, no symmetry assumption.

Once again — thank you for the diagnostic. This is exactly the right behavior. Closing #1567..
